EKF3: Improve stability of covariance matrix processing

priseborough opened the pull request:
1) Variances for each axis  are increased to prevent the ratio of the largest to smallest value exceeding 100. Large ratios can result in poor stability of the delta velocity bias state covariance prediction and update operations.

2) The covariance prediction step explicitly constrains the variance from falling below the target minimum of 1E-8 which is equivalent to about 0.01 m/sec^2 of accel bias uncertainty. 

These changes improve the protection against badly conditioned IMU delta velocity bias state variances. This is necessary due to the limitations associated with single precision arithmetic that limits the range of values that can be used for the EK3_ABIAS_P_NSE parameter. The order and rate that different sensors are fused changes the minimum safe value for this parameter meaning that the safe minimum parameter value varies between board types. This requires active protection measures.

The symptons of badly conditioned delta velocity bias variances are:

1) The accel bias variance for the sensor axis aligned with the vertical collapses to the minimum value allowed by the ConstrainVariances() function.
2) Delta velocity bias states  fail to converge on correct values and in extreme cases can even go to the limits imposed by the ConstrainStates()  function.

This is most likely to happen during static operation.

tridge commented:
thanks Paul! I'm testing this today

tridge commented:
first 2 test flights were good. 3rd one not so good:
http://uav.tridgell.net/tmp/000044.BIN
large attitude error, looks like accel biases went badly off

OXINARF commented on libraries/AP_NavEKF3/AP_NavEKF3_core.cpp:
```diff
@@ -863,15 +863,23 @@ void NavEKF3_core::CovariancePrediction()
+        // default process noise (m/s)^2
         float dVelBiasVar = sq(sq(dt) * constrain_float(frontend->_accelBiasProcessNoise, 0.0f, 1.0f));
+
+        // Find the maximum delta velocity bvias state variance
```
bvias -> bias

OXINARF commented on libraries/AP_NavEKF3/AP_NavEKF3_core.cpp:
```diff
@@ -863,15 +863,23 @@ void NavEKF3_core::CovariancePrediction()
+            }
+        }
+
+        // To ensure stability of the covaraince matrix operations, the ratio of a max and min variance must not exceed 100 and the minimum variance must not fall below 1E-8
```
covaraince -> covariance

priseborough commented:
The limiting of the variances was being done unnecessarily in two places so I have moved it all to the ConstrainVariances function and rebased. The default  state process noise parameter has also been increased. Here are the IMU delta velocity bias states comparison between EKF2 and EKF3 for the test. The Z bias learned is essentially identical:

![screen shot 2017-06-29 at 9 52 07 pm](https://user-images.githubusercontent.com/3596952/27686291-80475b2c-5d15-11e7-9315-0e9de897c170.png)

Here are the delta velocity bias state variances logged from EKF3. The delta velocity bias variances are being constrained to 1E-8 (roughly equivalent to 0.01 m/s/s bias uncertainty) as expected:

![screen shot 2017-06-29 at 9 55 49 pm](https://user-images.githubusercontent.com/3596952/27686389-e661db08-5d15-11e7-805b-8467fb8da5b0.png)

priseborough commented:
Depending on the results of further tests, I may increase the default value of EK3_ABIAS_P_NSE further  so that the process noise builds the variance up away from the lower constraint. The constraint is supposed to be there as protection  and not a substitute for correct tuning.

amilcarlucas commented:
Is anybody else testing this ?

priseborough commented:
@tridge That log you posted was from a version of the EKF that had a non-standard value set for the allowed and initial accel delta velocity bias uncertainty that the estimator was unable to handle.

amilcarlucas commented:
@priseborough is this WIP ? Should we add WIP label ?

rmackay9 commented:
@priseborough, if i ran some test flights with EKF3 before/after would that be sufficient to decide if this change is good or not?

rmackay9 commented:
Sorry, I haven't tested this yet but I will in the next 2 weeks.

priseborough commented:
That would be sufficient. The next step after this PR will be to implement the Kahan summation algorithm to work around the single precision floating point errors that prevent the process noise accumulating at the rate indicated by the EK3_ABIAS_P_NSE parameter.

priseborough commented:
I've shelved work on this until we get replay working with representative results. Otherwise it is going to be hard to tell if the proposed changes improve what is likely to be subtle change in the great majority of flights.

So far I haven't seen the instability in delta velocity bias estimates that this PR was created to prevent, but that could be because of a small number of EKF3 users.

peterbarker commented:
@priseborough I've got Replay back on its feet - even if it is, perhaps, wobbling slightly.

https://github.com/ArduPilot/ardupilot/pull/13885

It doesn't touch much outside Replay, so I'm hopeful we can get it merged soonish.

If there are ways I can test Replay to ensure they work for you, please let me know.

tridge commented:
@priseborough should we revive this one now?

priseborough commented:
Out of data now we have moved to the sympy auto-code. Might as well close it and i will create a new one using some updated knowledge on this issue.

priseborough commented:
I'm re-opening this as we now have replay working and which shows these changes fix the behaviour logged here: https://github.com/ArduPilot/ardupilot/issues/16791
